### DISCUSSION TITLE
Retrieving InlineKeyboardButton kwargs

### DISCUSSION THREAD

**Discussion by RezDaniel** [python-telegram-bot/python-telegram-bot#2269] [Q&A]
Hi everyone, I'm having a bit of trouble retrieving the kwargs sent from an InlineKeyboardButton?

For instance if this is my keyboard button...

```
keyboard = [
        [InlineKeyboardButton("1 hour", callback_data=str(D2), test=100)],
```

How do I then retrieve the kwarg value `100` from `test=100` and send it to a method?

**Comment by Poolitzer** [comment]
that should give you an error on the newest version. In older ones, it would be passed to telegrams api which disregards it and then you cant access it either

**Reply by RezDaniel** [reply]
Ahhh I see. In addition to `Callback_data`, is there a way to send InlineKeyboard Button args to a function once the button is pressed?

**Reply by Poolitzer** [reply]
Sure. As long as you handle the callback data + handlers correctly, you can use context.user_data. check out our wiki: https://github.com/python-telegram-bot/python-telegram-bot/wiki/Storing-bot,-user-and-chat-related-data

**Reply by Bibo-Joshi** [reply]
Just adding a bit more context: passing arbitrary kwargs to TG classes will not lead to errors, but they will simply be ignored. This is to ensure that API updates that add new kwargs to classes don't immediately break running bots.
Pool probably confused this with bot methods, where we no longer accept kwargs in favor of the (rather) new `api_kwargs` argument.

**Reply by RezDaniel** [reply]
Thank you both for your comments and help so far.  

Here's a snippet of my current code: https://www.codepile.net/pile/5v5b6QzP

What I am trying to do is refactor my code to reduce method repetition and only have one method that all buttons call back to while passing button specific data (but data I can pre-define). Something like this: https://www.codepile.net/pile/rVazkYQ7

Is this possible?

**Reply by Bibo-Joshi** [reply]
you could simply pass the `buy_amount` as `callback_data` and retrieve it from the pressed button. Than you would only need one callback method. As a side note: You shouldn't use module wide variables like `boy_amount`. Instead rather use `user_data` as described in the link Poolitzer posted above

**Reply by RezDaniel** [reply]
How do I set the `callback_data`? When I use the below definitions...

`BUY0, BUY50, BUY250, BUY1000, BUY5000) = range(5)` 

...and retrieve using `update.callback_query.data`, the data is returned `BUY0 = 0, BUY50 = 1, BUY250 = 2, BUY1000 = 3, BUY5000 = 4` as consecutive ints depending on the order there defined, rather than the actual string ("BUY0" etc.) being return?

Thanks for the heads up about the module wide vars, I'll use the `user_data` dict moving forward.

**Reply by Bibo-Joshi** [reply]
you can just pass `callback_data='500'`. the conversation states don't necessarily have anything to do with the callback data.

**Reply by RezDaniel** [reply]
Right, I understand now. Perfect, thank you both for your detailed input and fast responses, very much appreciated, reduced my code base by over 50%.
